apps/src/p5lab/redux/animationList.js

```diff
@@ -260,10 +259,12 @@ function generateAnimationName(baseName, animationList) {
 
 /**
  * @param {!SerializedAnimationList} serializedAnimationList
+ * @param {Object} alternativeDefaultSprites - optional list of default sprites. Otherwise, use serializedAnimationList
```

Comment:
This is where I'd really like feedback. This function is called four times - twice in P5Lab and once each in blocks/edit and blocks/index. In three of those places, the serializedAnimationList passed in here is the same as what is fetched as defaultSprites. However, on line 261 in P5Lab.js, I can't guarantee that the serializedAnimationList is equal to the defaultSprites, so I added alternativeDefaultSprites as a way to specify if there is an alternative list we should be using for defaultSprites and then made it optional and set the default value to serializedAnimationList, so that in 3 of the 4 cases, we didn't need to pass in anything. Thoughts?

Reply:
since `alternativeDefaultSprites` is only used in one case -- the one we're adding here -- i'd do this work in P5Lab instead to merge serializedAnimationList and defaultSprites before we call this method

Reply:
I'm not following what you are suggesting. I played around with this a bit before landing on this solution. One option, is to just pass in defaultSprites in all cases, but that ended up with 3 of the 4 calls being something like, `setInitialAnimationList(defaultSprites, true, false, defaultSprites)`. Does that align with what you were thinking? It would mean that the defaultSprites here wouldn't need an optional parameter.

Reply:
no, i mean remove the `alternativeDefaultSprites` argument. then, in P5Lab.js, we can merge `initialAnimationList` and `defaultSprites` before calling `setInitialAnimationList`:

```javascript
// P5Lab.js

// below is just an example -- this likely isn't actually what we want to do.
// combine initialAnimationList & defaultSprites to create the serializedAnimationList
// required by setInitialAnimationList
const serializedAnimationList = {...initialAnimationList, ...defaultSprites};

getStore().dispatch(
    setInitialAnimationList(
      serializedAnimationList,
      this.isSpritelab /* shouldRunV3Migration */,
      this.isSpritelab
    )
);
```

Reply:
Hmm, maybe I'm not understanding, but I'm not sure that would accomplish what we are wanting it to. The defaultSprites in animationList are used to replace the values in the serializedAnimationList, not necessarily combine with them, so the code above could result in adding extra animations to the initialAnimationList than what was intended. In animationList, the defaultSprites are also only used in a very specific migration case (`if(shouldRunV3Migration) and if(animation.sourceUrl.includes('/v3/')) and if(defaultSprites.propsByKey[loadedKey])`). With what we know about usage from Google Analytics (https://codedotorg.slack.com/archives/C1B3PNDL7/p1629153431047500), we know this case doesn't account for all the usage here, so there are times where we do not want the serializedAnimationList values to be replaced by the defaultSprite values. If we wanted to remove the defaultSprites as a parameter in setInitialAnimationList, without adding a promise structure to setInitialAnimationList (which would have other side effects), we could pull the migration out of setInitialAnimationList and perform it in P5Lab in the place you suggested above. However, since there are two separate migrations here, I would probably push back on moving one (the 2020 one), without moving the other (the 2015 one). So we could move the two migrations out of setInitialAnimationList. I also think the migrations make more sense to have in the setInitialAnimationList than in P5Lab from a structural point of view, although if you feel strongly, I could be convinced to move them.

Reply:
Hmm, maybe I'm not understanding, but I'm not sure I agree with this move. In animationList, the defaultSprites are also only used in a very specific migration case (if(shouldRunV3Migration) and if(animation.sourceUrl.includes('/v3/')) and if(defaultSprites.propsByKey[loadedKey])). With what we know about usage from Google Analytics (https://codedotorg.slack.com/archives/C1B3PNDL7/p1629153431047500), we know this case doesn't account for all the usage here, so there are times where we do not want the serializedAnimationList values to be replaced by the defaultSprite values. If we move the whole 2020 migration, to account for this merging in P5Lab, I would want to bring both migrations (the 2020 and the 2015), since I think they make more sense in context together. Overall, I think the migrations make more sense to have in animationList than in P5Lab, since they are tied pretty closely with the context of the initial animationList, although if you feel strongly, I could be convinced to move them.

Reply:
ah, i see -- i didn't understand that piece before. i don't really see an alternative route than the one you're taking here, in that case. the issue is that this code (even just the `setInitialAnimationList` function) has become increasingly complex, which is why you're having to work around the existing functionality in this way.

for now, we should add test coverage for this new logic and make sure the existing test coverage for this function is sufficient. then, i think we should log tasks to remove the migrations whenever we can, and refactor away the complexity in this function once those are removed.

Reply:
Yup, we have a ticket to fix the 2020 migration, but I'll make sure to note that if possible, we should remove the defaultSprites (seems like we might need to keep it a little longer from the Slack thread): https://codedotorg.atlassian.net/browse/STAR-1687.

I added a test to cover the migration branch.